File: scenarios/calibration/python-sales-by-genre/tests/apply_patch.py
#!/usr/bin/env python3
"""Apply the trusted single-file unified patch to a verifier-only baseline."""

from __future__ import annotations

import re
import sys
from pathlib import Path


HUNK = re.compile(r"^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@")


def apply(original: str, patch: str) -> str:
    source = original.splitlines(keepends=True)
    patch_lines = patch.splitlines(keepends=True)
    output: list[str] = []
    source_index = 0
    index = 0
    while index < len(patch_lines):
        match = HUNK.match(patch_lines[index])
        if match is None:
            index += 1
            continue
        old_start = int(match.group(1)) - 1
        if old_start < source_index:
            raise ValueError("overlapping patch hunks")
        output.extend(source[source_index:old_start])
        source_index = old_start
        index += 1
        while index < len(patch_lines) and not patch_lines[index].startswith("@@ ") and not patch_lines[index].startswith("diff --git "):
            line = patch_lines[index]
            if line.startswith("\\ No newline at end of file"):
                index += 1
                continue
            marker = line[:1]
            content = line[1:]
            if marker in {" ", "-"}:
                if source_index >= len(source) or source[source_index] != content:
                    raise ValueError("patch context does not match baseline")
                if marker == " ":
                    output.append(content)
                source_index += 1
            elif marker == "+":
                output.append(content)
            else:
                break
            index += 1
    output.extend(source[source_index:])
    return "".join(output)


def main() -> None:
    target = Path(sys.argv[1])
    patch = Path(sys.argv[2]).read_text(encoding="utf-8")
    if patch:
        target.write_text(apply(target.read_text(encoding="utf-8"), patch), encoding="utf-8")


if __name__ == "__main__":
    main()

File: scenarios/calibration/react-author-filter/tests/apply-patch.mjs
#!/usr/bin/env node
import fs from 'node:fs';

const [, , targetPath, patchPath] = process.argv;
const original = fs.readFileSync(targetPath, 'utf8');
const patch = fs.readFileSync(patchPath, 'utf8');
if (patch.length === 0) process.exit(0);

const source = original.match(/[^\n]*\n|[^\n]+$/g) ?? [];
const lines = patch.match(/[^\n]*\n|[^\n]+$/g) ?? [];
const output = [];
let sourceIndex = 0;
let index = 0;
while (index < lines.length) {
  const match = /^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@/.exec(lines[index]);
  if (!match) {
    index += 1;
    continue;
  }
  const oldStart = Number(match[1]) - 1;
  if (oldStart < sourceIndex) throw new Error('overlapping patch hunks');
  output.push(...source.slice(sourceIndex, oldStart));
  sourceIndex = oldStart;
  index += 1;
  while (index < lines.length && !lines[index].startsWith('@@ ') && !lines[index].startsWith('diff --git ')) {
    const line = lines[index];
    if (line.startsWith('\\ No newline at end of file')) {
      index += 1;
      continue;
    }
    const marker = line[0];
    const content = line.slice(1);
    if (marker === ' ' || marker === '-') {
      if (source[sourceIndex] !== content) throw new Error('patch context does not match baseline');
      if (marker === ' ') output.push(content);
      sourceIndex += 1;
    } else if (marker === '+') {
      output.push(content);
    } else {
      break;
    }
    index += 1;
  }
}
output.push(...source.slice(sourceIndex));
fs.writeFileSync(targetPath, output.join(''));
